File: Libs/TauMPC/src/mpc/circuit/Circuit.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package mpc.circuit;

import mpc.compiler.Parser;
import mpc.finite_field_math.Zp;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import mpc.circuit.Gate.Operation;


public class Circuit {
    private static Map<Integer, Circuit> exponentialGates = new HashMap<Integer, Circuit>();
    protected List<Gate> gates;
    protected List<String> inputs;
    protected int p;

    public Circuit(int p, List<Gate> gates, List<String> inputs) {
        this.p = p;
        this.gates =  gates;
        this.inputs = inputs;
    }

    public void setGates(List<Gate> gates) {
        this.gates = gates;
    }        
    
    public List<Gate> getGates() {
        assert gates!= null : "run calculateGates first and check output";
        return gates;
    }

    public void setInputs(List<String> inputs) {
        this.inputs = inputs;
    }

    public List<String> getInputs() {
        return inputs;
    }        
    
    //returns a map of all outputs (the int is the output index)
    public Map<Integer, Zp> getOutputs() {
        Map<Integer, Zp> result = new HashMap<Integer, Zp>();
        for (Gate gate : gates) {
            for (Wire wire : gate.getOutputWires()) {
                if (wire.getOutputIndex() != null) {
                    assert gate.isOutputReady();
                    result.put(wire.getOutputIndex(), gate.getOutputValue());
                }
            }
        }
        return result;
    }
    
    //returns the number of inputs declared in code - not all inputs must be used in curcuit
    public int getCircuitInputSize(){
        return inputs.size();        
    }
    
    //returns the number of actual inputs - that are used in the circuit
    public int getCircuitRealInputSize(){
        List<Wire> inputWires = getInputWires();
        Set<Integer> currInputs = new HashSet<Integer>();
        for (Wire wire : inputWires){
            if (wire.getInputIndex() != null){
                currInputs.add(wire.getInputIndex());
            }
        }
        return currInputs.size();        
    }
    
    public boolean isMultipleContained(){
        for (Gate gate : gates){
            if (gate.getOperation() == Operation.MUL || gate.getOperation() == Gate.Operation.DIV){
                return true;
            }
        }
        return false;
    }
    
    public List<Wire> getInputWires(){
        return getInputOrOutputWires(true);
    }
    
    public List<Wire> getOutputWires(){
        return getInputOrOutputWires(false);
    }
    
    private List<Wire> getInputOrOutputWires(final boolean isInput){//todo - check that there are no duplicates
        List<Wire> inputOrOutputWires = new ArrayList<Wire>();
        for (Gate gate : gates){
            List<Wire> wires = isInput ? gate.getInputWires() : gate.getOutputWires();
            for (Wire wire : wires){
                if (wire.isInput() && isInput || wire.isOutput() && !isInput){
                    inputOrOutputWires.add(wire);
                }
            }
        }
        Collections.sort(inputOrOutputWires, new Comparator(){
            public int compare(Object o1, Object o2) {
                Wire w1 = (Wire)o1;
                Wire w2 = (Wire)o2;
                Integer v1 = isInput ? w1.getInputIndex() : w1.getOutputIndex();
                Integer v2 = isInput ? w2.getInputIndex() : w2.getOutputIndex();
                return v1.compareTo(v2);
            }                        
        });
        return inputOrOutputWires;
    }

    public void emptyResult(){
        for (Gate gate : gates){
            gate.deleteOutputValue();
        }
    }        
            
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Inputs: ");
        List<Wire> inputWires = getInputWires();
        List<Integer> usedInputs = new ArrayList<Integer>();
        StringBuilder inputWiresSB = new StringBuilder();
        for (Wire wire : inputWires) {
            if (usedInputs.contains(wire.getInputIndex())) {
                continue;
            }
            inputWiresSB.append(wire.getInputIndex()).append(", ");
            usedInputs.add(wire.getInputIndex());
        }
        sb.append(inputWiresSB.toString().substring(0, inputWiresSB.length() - 2)).append("\n");
        for (Gate gate : gates){
            sb.append("Gate").append(gates.indexOf(gate)).append(": ").append(gate.display(gates)).append("\n");
        }
        sb.append("Outputs: ");
        List<Wire> outputWires = getOutputWires();
        for (Wire wire : outputWires){
            sb.append("Gate").append(gates.indexOf(wire.getSourceGate()));         
            if (outputWires.indexOf(wire) < outputWires.size() - 1){
                sb.append(", ");
            }
        }
        return sb.toString();
    }
    
    public static Circuit getDivisionExponentialCircuit(int p){
        Circuit c = exponentialGates.get(p);
        if (c == null){            
            c = new ExponentCircuit(p, p - 2).getCircuit();
            if (c == null){
                c = createExponentCircuit(p - 2, p);                
            }          
            exponentialGates.put(p, c);
            return c;
        }
        c.emptyResult();
        return c;
    }
    
    private static class ExponentCircuit{
        private int prime;
        private int exp;
        private Map<Integer, Wire> calculatedValues = new HashMap<Integer, Wire>();

        public ExponentCircuit(int prime, int exp) {
            this.prime = prime;
            this.exp = exp;
        }
        
        public Circuit getCircuit(){
            Wire outputWire = getWireForExp(exp);
            outputWire.setOutputIndex(0);
            try {
                return Parser.createCircuit(Arrays.asList(outputWire), prime, Arrays.asList("x"));
            } catch (ParseException ex) {
                assert false;//should not happen
                return null;
            }                        
        }
        
        private Wire getWireForExp(int exp){
            Wire wire = calculatedValues.get(exp);
            if (wire != null){
                return wire.clone();
            }          
            if (exp == 1){
                return new Wire(0, true);
            }
            int biggestExponential = getBiggest2Exponential(exp);
            if (biggestExponential == exp){
                return getWireFor2ExponentialExp(exp);                
            }
            Wire firstWire = getWireForExp(biggestExponential);
            Wire secondWire = getWireForExp(exp - biggestExponential);
            return createGate(firstWire, secondWire);            
        }
        
        private Wire getWireFor2ExponentialExp(int exp){
            int numberOfIterations = get2Log(exp);
            Wire firstWire = new Wire(0, true);
            Wire secondWire = new Wire(0, true);      
            Wire outputWire = createGate(firstWire, secondWire);
            calculatedValues.put(2, outputWire);
            for (int i = 1; i < numberOfIterations; i++) {
                firstWire = outputWire;
                secondWire = outputWire.clone();
                outputWire = createGate(firstWire, secondWire);
                calculatedValues.put((int)Math.pow(2, i + 1), outputWire);
            }
            calculatedValues.put(exp, outputWire);
            return outputWire;            
        }
        
        private Wire createGate(Wire firstWire, Wire secondWire) {
            Wire outputWire = new Wire();
            Gate gate = new Gate(Arrays.asList(firstWire, secondWire), outputWire, Operation.MUL, prime);
            firstWire.setTargetGate(gate);
            secondWire.setTargetGate(gate);
            outputWire.setSourceGate(gate);
            return outputWire;
        }
        
        private int getBiggest2Exponential(int max){
            if (max == 0){
                throw new RuntimeException("Cannot find 2 exponential");
            }            
            int i = 1;
            for (i = 1; i <= max; i = 2 * i){}
            return i/2;
        }
        
        private int get2Log(int num){
            assert num == getBiggest2Exponential(num);
            int j = 0;
            for (int i = 1; i <= num; i = 2 * i){
                j++;
            }
            return j - 1;
        }
        
    }
    
    public static Circuit createExponentCircuit(int exp, int p){
        List<Wire> currWires = new ArrayList<Wire>();
        for (int i = 0; i < exp; i++){
            currWires.add(new Wire(0, true));
        }
        List<Gate> gates = new ArrayList<Gate>();
        while(currWires.size() > 1){
            List<Wire> outputWires = new ArrayList<Wire>();
            for (int i = 0; i < currWires.size() - 1; i = i + 2){
                Wire outputWire = new Wire();
                Gate gate = new Gate(Arrays.asList(currWires.get(i), currWires.get(i + 1)), outputWire, Operation.MUL, p);
                currWires.get(i).setTargetGate(gate);
                currWires.get(i + 1).setTargetGate(gate);
                outputWire.setSourceGate(gate);
                outputWires.add(outputWire);
                gates.add(gate);
            }
            if (currWires.size() % 2 != 0){
                outputWires.add(currWires.get(currWires.size() - 1));
            }
            currWires = outputWires;
        }
        currWires.get(0).setOutputIndex(0);
        return new Circuit(p, gates, Arrays.asList("x"));
    }

    //returns a List of all outputs - for internal use only
    public  List<Zp>  internalCalculate(String inputs, int prime)  {

        inputs = inputs.replaceAll(" ", "");//remove white spaces
        String[] tokens = inputs.split(",");
        List<Zp> usersInputs = new ArrayList<Zp>();
        for (int i=0; i< tokens.length; i++) {
              usersInputs.add( new Zp(prime, Integer.parseInt(tokens[i])));
         }

        for (Gate gate : getGates()){
            gate.internalCalculate(usersInputs);
        }

        Map<Integer, Zp> outputsMap = getOutputs();

        int numOfOutputs = outputsMap.size();
        Zp tempRes = null;
        List<Zp> resultList = new ArrayList<Zp>();

        for (int i = 0; i < numOfOutputs; i++){
            tempRes = outputsMap.get(new Integer(i));
            resultList.add(tempRes);
        }
        return resultList;
    }
    
    public static void main(String[] args){
        Circuit circuit = Circuit.getDivisionExponentialCircuit(61);
        System.out.print(circuit.toString());        
    }
  }

File: Libs/FiniteField/Circuits/Circuit.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text.RegularExpressions;

namespace MpcLib.Common.FiniteField.Circuits
{
	public class Circuit
	{
		//Dictionary<int, Circuit> exponentialGates = new Dictionary<int, Circuit>();
		public IList<Gate> Gates { get; set; }

		public IList<string> Inputs { get; set; }

		public virtual IList<Wire> InputWires
		{
			get
			{
				return GetInputOrOutputWires(true);
			}
		}

		public readonly Gate OutputGate;
		protected internal int Prime;

		public Circuit(int prime, IList<Gate> gates, IList<string> inputs, Gate outputGate)
		{
			Prime = prime;
			Gates = gates;
			Inputs = inputs;
			OutputGate = outputGate;
		}

		public virtual Zp Output
		{
			get
			{
				foreach (var gate in Gates)
				{
					foreach (var wire in gate.OutputWires)
					{
						if (wire.IsOutput)
						{
							Debug.Assert(gate.IsOutputReady);
							return gate.OutputValue;
						}
					}
				}
				return null;
			}
		}

		/// <summary>
		/// The number of inputs declared in code - not all inputs must be used in curcuit
		/// </summary>
		public virtual int InputCount
		{
			get
			{
				return Inputs.Count;
			}
		}

		// Commented by Mahdi: Just do not need!
		///// <summary>
		///// The number of actual inputs - that are used in the circuit
		///// </summary>
		//public virtual int RealInputCount
		//{
		//    get
		//    {
		//        IList<Wire> inputWires = InputWires;
		//        var currInputs = new HashSet<int>();
		//        foreach (Wire wire in inputWires)
		//        {
		//            if (wire.InputIndex != null)
		//                currInputs.Add(wire.InputIndex);
		//        }
		//        return currInputs.Count;
		//    }
		//}

		public virtual bool MultipleContained
		{
			get
			{
				foreach (Gate gate in Gates)
				{
					if (gate.Operation == Operation.Mul || gate.Operation == Operation.Div)
						return true;
				}
				return false;
			}
		}

		private IList<Wire> GetInputOrOutputWires(bool isInput) //todo - check that there are no duplicates
		{
			var inputOrOutputWires = new List<Wire>();
			foreach (Gate gate in Gates)
			{
				var wires = isInput ? gate.InputWires : gate.OutputWires;
				foreach (Wire wire in wires)
				{
					if (wire.IsInput && isInput || wire.IsOutput && !isInput)
						inputOrOutputWires.Add(wire);
				}
			}
			return inputOrOutputWires;
		}

		public virtual void EmptyResult()
		{
			foreach (Gate gate in Gates)
				gate.DeleteOutputValue();
		}

		public override string ToString()
		{
			var sb = "Inputs: ";
			var inputWires = InputWires;
			var usedInputs = new List<int>();
			var inputWiresSB = "";

			foreach (Wire wire in inputWires)
			{
				if (usedInputs.Contains(wire.InputIndex))
					continue;

				inputWiresSB += wire.InputIndex + ", ";
				usedInputs.Add(wire.InputIndex);
			}
			sb += inputWiresSB.Substring(0, inputWiresSB.Length - 2) + " | ";
			foreach (Gate gate in Gates)
				sb += "Gate" + Gates.IndexOf(gate) + ": " + gate.Display(Gates) + " | ";

			sb += "Output: Gate" + Gates.IndexOf(OutputGate);
			return sb.ToString();
		}

		//public static Circuit CreateExponentCircuit(int exp, int p)
		//{
		//    var currWires = new List<Wire>();
		//    for (int i = 0; i < exp; i++)
		//        currWires.Add(new Wire(0, true));

		//    var gates = new List<Gate>();
		//    while (currWires.Count > 1)
		//    {
		//        var outputWires = new List<Wire>();
		//        for (int i = 0; i < currWires.Count - 1; i = i + 2)
		//        {
		//            var outputWire = new Wire();
		//            Gate gate = new Gate(new List<Wire>(){ currWires[i], currWires[i + 1] },
		//                outputWire, Operation.Mul, p);

		//            currWires[i].TargetGate = gate;
		//            currWires[i + 1].TargetGate = gate;
		//            outputWire.SourceGate = gate;
		//            outputWires.Add(outputWire);
		//            gates.Add(gate);
		//        }
		//        if (currWires.Count % 2 != 0)
		//            outputWires.Add(currWires[currWires.Count - 1]);

		//        currWires = outputWires;
		//    }
		//    currWires[0].IsOutput = 0;
		//    return new Circuit(p, gates, new List<string>() { "x" });
		//}

		private static string[] StringSplit(string source, string regexDelimiter, bool trimTrailingEmptyStrings)
		{
			var splitArray = Regex.Split(source, regexDelimiter);

			if (trimTrailingEmptyStrings)
			{
				if (splitArray.Length > 1)
				{
					for (int i = splitArray.Length; i > 0; i--)
					{
						if (splitArray[i - 1].Length > 0)
						{
							if (i < splitArray.Length)
								System.Array.Resize(ref splitArray, i);
							break;
						}
					}
				}
			}
			return splitArray;
		}

		/// <summary>
		/// Returns a List of all outputs.
		/// </summary>
		internal Zp InternalCalculate(string inputs, int prime)
		{
			inputs = inputs.Replace(" ", "");	// remove white spaces
			string[] tokens = StringSplit(inputs, ",", true);
			var usersInputs = new List<Zp>();

			for (int i = 0; i < tokens.Length; i++)
				usersInputs.Add(new Zp(prime, Convert.ToInt32(tokens[i])));

			foreach (Gate gate in Gates)
				gate.InternalCalculate(usersInputs);

			return Output;
		}

		//public Circuit getDivisionExponentialCircuit(int p)
		//{
		//    var c = exponentialGates[p];
		//    if (c == null)
		//    {
		//        c = (new ExponentCircuit(p, p - 2)).Circuit;
		//        if (c == null)
		//            c = createExponentCircuit(p - 2, p);

		//        exponentialGates[p] = c;
		//        return c;
		//    }
		//    c.emptyResult();
		//    return c;
		//}

		//private class ExponentCircuit
		//{
		//    private int prime;
		//    private int exp;
		//    Dictionary<int, Wire> calculatedValues = new Dictionary<int, Wire>();

		//    public ExponentCircuit(int prime, int exp)
		//    {
		//        this.prime = prime;
		//        this.exp = exp;
		//    }

		//    internal virtual Circuit Circuit
		//    {
		//        get
		//        {
		//            var outputWire = getWireForExp(exp);
		//            outputWire.IsOutput = 0;
		//            try
		//            {
		//                return Parser.CreateCircuit(new List<Wire>() { outputWire }, prime, new List<string>(){ "x" });
		//            }
		//            catch (Exception)
		//            {
		//                Debug.Assert(false);
		//                return null;
		//            }
		//        }
		//    }

		//    private Wire getWireForExp(int exp)
		//    {
		//        Wire wire = calculatedValues[exp];
		//        if (wire != null)
		//            return wire.Clone();

		//        if (exp == 1)
		//            return new Wire(0, true);

		//        int biggestExponential = getBiggest2Exponential(exp);
		//        if (biggestExponential == exp)
		//        {
		//            return getWireFor2ExponentialExp(exp);
		//        }
		//        Wire firstWire = getWireForExp(biggestExponential);
		//        Wire secondWire = getWireForExp(exp - biggestExponential);
		//        return createGate(firstWire, secondWire);
		//    }

		//    private Wire getWireFor2ExponentialExp(int exp)
		//    {
		//        int numberOfIterations = get2Log(exp);
		//        var firstWire = new Wire(0, true);
		//        var secondWire = new Wire(0, true);
		//        Wire outputWire = createGate(firstWire, secondWire);
		//        calculatedValues[2] = outputWire;
		//        for (int i = 1; i < numberOfIterations; i++)
		//        {
		//            firstWire = outputWire;
		//            secondWire = outputWire.Clone();
		//            outputWire = createGate(firstWire, secondWire);
		//            calculatedValues[(int)Math.Pow(2, i + 1)] = outputWire;
		//        }
		//        calculatedValues[exp] = outputWire;
		//        return outputWire;
		//    }

		//    private Wire createGate(Wire firstWire, Wire secondWire)
		//    {
		//        var outputWire = new Wire();
		//        var gate = new Gate(
		//            new List<Wire>() { firstWire, secondWire }, outputWire, Operation.Mul, prime);
		//        firstWire.TargetGate = gate;
		//        secondWire.TargetGate = gate;
		//        outputWire.SourceGate = gate;
		//        return outputWire;
		//    }

		//    private int getBiggest2Exponential(int max)
		//    {
		//        if (max == 0)
		//            throw new Exception("Cannot find 2 exponential");

		//        int i = 1;
		//        for (i = 1; i <= max; i = 2 * i) ;
		//        return i / 2;
		//    }

		//    private int get2Log(int num)
		//    {
		//        Debug.Assert(num == getBiggest2Exponential(num));
		//        int j = 0;
		//        for (int i = 1; i <= num; i = 2 * i) j++;
		//        return j - 1;
		//    }
		//}
	}
}